homeassistant/components/google_assistant/trait.py

```diff
@@ -609,6 +614,102 @@ async def execute(self, command, data, params, challenge):
         )
 
 
+class EnergyStorageTrait(_Trait):
+    """Trait to offer EnergyStorage functionality.
+
+    https://developers.google.com/actions/smarthome/traits/energystorage
+    """
+
+    name = TRAIT_ENERGYSTORAGE
+    commands = [COMMAND_CHARGE]
+
+    @staticmethod
+    def supported(domain, features, device_class, _):
+        """Test if state is supported."""
+        return domain == vacuum.DOMAIN and features & vacuum.SUPPORT_BATTERY
+
+    def sync_attributes(self):
+        """Return EnergyStorage attributes for a sync request."""
+        return {
+            "isRechargeable": True,
+            "queryOnlyEnergyStorage": False,
+        }
+
+    def query_attributes(self):
+        """Return EnergyStorage query attributes."""
+        battery_level = self.state.attributes.get(ATTR_BATTERY_LEVEL)
+        if battery_level == 100:
+            descriptive_capacity_remaining = "FULL"
+        elif 75 <= battery_level < 100:
+            descriptive_capacity_remaining = "HIGH"
+        elif 50 <= battery_level < 75:
+            descriptive_capacity_remaining = "MEDIUM"
+        elif 25 <= battery_level < 50:
+            descriptive_capacity_remaining = "LOW"
+        elif 0 <= battery_level < 25:
+            descriptive_capacity_remaining = "CRITICALLY_LOW"
+        return {
+            "descriptiveCapacityRemaining": descriptive_capacity_remaining,
+            "capacityRemaining": [{"rawValue": battery_level, "unit": "PERCENTAGE"}],
+            "capacityUntilFull": [
+                {"rawValue": 100 - battery_level, "unit": "PERCENTAGE"}
+            ],
+            "isCharging": self.state.state == vacuum.STATE_DOCKED,
+            "isPluggedIn": self.state.state == vacuum.STATE_DOCKED,
+        }
+
+    async def execute(self, command, data, params, challenge):
+        """Execute a dock command to start charging."""
+        if params["charge"]:
+            if self.state.state == vacuum.STATE_DOCKED:
+                raise SmartHomeError(
+                    ERR_ALREADY_DOCKED,
+                    "The vacuum is already docked",
+                )
+            else:
+                await self.hass.services.async_call(
+                    self.state.domain,
+                    vacuum.SERVICE_RETURN_TO_BASE,
+                    {ATTR_ENTITY_ID: self.state.entity_id},
+                    blocking=True,
+                    context=data.context,
+                )
+        else:
+            if self.state.state == vacuum.STATE_DOCKED:
+                if (
+                    self.state.attributes.get(ATTR_SUPPORTED_FEATURES, 0)
+                    & vacuum.SUPPORT_PAUSE
+                ):
+                    await self.hass.services.async_call(
+                        self.state.domain,
+                        vacuum.SERVICE_START_PAUSE,
+                        {ATTR_ENTITY_ID: self.state.entity_id},
+                        blocking=True,
+                        context=data.context,
+                    )
```

Comment:
What effect is this expected to have? Seems like some toggle service (didn't even find it in dev docs). Seems it could either start cleaning or pause.

Reply:
vacuums that are docked are already in PAUSE or STOPPED state, so that toggle will either be an UNPAUSE or START, either resuming an existing cleaning or starting a full one

Reply:
now irrelevant